### PR TITLE
Disable DML EP on software adapter, fix float16 fallback bug, re-enable DML in CI

### PR DESCRIPTION
**Pull request by jeffbloo**
DML was previously disabled in the CI when VMs were transitioned.  It's believed the failures and timeouts were caused by the software adapter being temporarily used.

Private runs don't reproduce the issue, but did hit a couple of new FP16 model failures.  Using the older GPU appears to expose a DML EP bug with float16 fallback for some models (the fp16 tests are already disabled on CUDA).

This change:
1) Fixes the issue with float16 fallback by using the correct logic to predict partitioning when deciding whether to do DML graph fusions.  Note, the whole approach of doing graph transforms prior to partitioning is temporary.  This will eventually be done during creation of the partition kernel on a mutable subgraph.
2) Disables DML EP from being used on the software adapter, like Winml.  It's much faster to use the CPU EP.
3) Re-enables DML in the CI

### CODE REVIEW

File: onnxruntime/core/providers/dml/dml_provider_factory.cc [hunk 1]
```diff
@@ -49,13 +51,30 @@ std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_DML(ID
...
+
 std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_DML(int device_id) {
   ComPtr<IDXGIFactory4> dxgi_factory;
   THROW_IF_FAILED(CreateDXGIFactory2(0, IID_PPV_ARGS(&dxgi_factory)));
 
   ComPtr<IDXGIAdapter1> adapter;
   THROW_IF_FAILED(dxgi_factory->EnumAdapters1(device_id, &adapter));
 
+  // Disallow using DML with the software adapter (Microsoft Basic Display Adapter) because CPU evaluations are much faster
+  THROW_HR_IF(E_INVALIDARG, IsSoftwareAdapter(adapter.Get()));
```
Comment (snnn) [inline onnxruntime/core/providers/dml/dml_provider_factory.cc]: In Windows "Device Manager", I saw each our build machine has two display adapters:
1. The tesla
2. A basic display adapter

I guess both of them will be returned from the "EnumAdapters1" function?
Is it possible that sometimes the device of id 0 is the tesla but sometimes it is the basic one?

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.h [hunk 2]
```diff
@@ -158,6 +158,9 @@ namespace Dml
         std::shared_ptr<onnxruntime::IAllocator> GetCpuInputAllocator();
         std::shared_ptr<onnxruntime::IAllocator> GetCpuOutputAllocator();
 
+        std::shared_ptr<const winrt::Windows::AI::MachineLearning::implementation::InternalRegistrationInfoMap> 
+        GetInternalRegInfoMap() const;
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.h]: >Reg [](start = 19, length = 3)

`Registration` preferred.

Comment (fdwr) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.h]: (though I see `m_internalRegInfoMap` is already abbreviate..)

---
In reply to: [369979610](https://github.com/microsoft/onnxruntime/pull/2896#discussion_r369979610) [](ancestors = 369979610)

File: onnxruntime/core/providers/dml/dml_provider_factory.cc [hunk 3]
```diff
@@ -49,13 +51,30 @@ std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_DML(ID
   return std::make_shared<onnxruntime::DMLProviderFactory>(dml_device, cmd_queue);
 }
 
+bool IsSoftwareAdapter(IDXGIAdapter1* adapter)
+{
+    DXGI_ADAPTER_DESC1 pDesc;
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/dml_provider_factory.cc]: >pDesc [](start = 23, length = 5)

`p`? Is this really a pointer to anything (and even if it is, didn't we kill Hungarian)?

Propose `adapterDesc`.

File: onnxruntime/core/providers/dml/dml_provider_factory.cc [hunk 4]
```diff
@@ -91,13 +110,17 @@ std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_DML(in
 }  // namespace onnxruntime
 
 ORT_API_STATUS_IMPL(OrtSessionOptionsAppendExecutionProvider_DML, _In_ OrtSessionOptions* options, int device_id) {
+API_IMPL_BEGIN
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/dml_provider_factory.cc]: [minor] Consistent indentation.
(sheesh, indentation styles in ORT seems so wonky and inconsistent)

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphPartitioner.cpp [hunk 5]
```diff
@@ -219,11 +219,14 @@ namespace Dml
 
     bool DoesNodeContainSupportedDataTypes(
         const onnxruntime::Node& node,
-        const std::unordered_map<std::string, GraphPartition*>& nodeNameToPartitionMap,
+        bool allow64BitInputThroughStrides,
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphPartitioner.cpp]: >allow64BitInputThroughStrides [](start = 13, length = 29)

o_O
Oh boy, this new parameter rather complicates already complex code. I'm looking forward to hopefully deleting it later with [https://microsoft.visualstudio.com/OS/_workitems/edit/24643526](https://microsoft.visualstudio.com/OS/_workitems/edit/24643526).

File: onnxruntime/core/providers/dml/dml_provider_factory.cc (hunk 1 above)
Comment (jeffbloo) [reply onnxruntime/core/providers/dml/dml_provider_factory.cc]: The hardware device will be enumerated first

File: onnxruntime/core/providers/dml/dml_provider_factory.cc (hunk 4 above)
Comment (jeffbloo) [reply onnxruntime/core/providers/dml/dml_provider_factory.cc]: The ORT code is auto-formatted, so the indents wouldn't survive...

Comment (fdwr) [reply onnxruntime/core/providers/dml/dml_provider_factory.cc]: Ugh, auto formatters are the bane of readability.

---
In reply to: [370344645](https://github.com/microsoft/onnxruntime/pull/2896#discussion_r370344645) [](ancestors = 370344645)

**Pull request merged**